readme.md
# Barzahlen Payment Module (Prestashop 1.4)

## User Manual
https://integration.barzahlen.de/de/shopsysteme/prestashop/nutzerhandbuch-14

## Current Version
1.0.2

## Changelog

### 1.0.2 (25.06.2013)
* updated Barzahlen API PHP SDK
* added automatic payment slip cancelation on order cancelation
* implemented automatic plugin version check (weekly)
* using PSR2 coding standard

### 1.0.1 (11.03.2013)
* changed image urls to make use of https
* updated cURL certificate bundle

### 1.0.0 (06.03.2013)
* initial release

## Support
The Barzahlen Team will happily assist you with any problems or questions. Send us an email to support@barzahlen.de or use the contact form at http://www.barzahlen.de/partner/integration.

## Copyright
(c) 2014, Cash Payment Solutions GmbH  
https://www.barzahlen.de

## NOTICE OF LICENSE
This source file is subject to the Open Software License (OSL 3.0) that is bundled with this package in the file LICENSE.txt.

It is also available through the world-wide-web at this URL: http://opensource.org/licenses/osl-3.0.php

If you did not receive a copy of the license and are unable to obtain it through the world-wide-web, please send an email to info@barzahlen.de so we can send you a copy immediately.

changelog.txt
==1.0.2 (25.06.2013)==
* updated Barzahlen API PHP SDK
* added automatic payment slip cancelation on order cancelation
* implemented automatic plugin version check (weekly)
* using PSR2 coding standard

==1.0.1 (11.03.2013)==
* changed image urls to make use of https
* updated cURL certificate bundle

==1.0.0 (06.03.2013)==
* initial release

src/modules/barzahlen/de.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

global $_MODULE;
$_MODULE = array();
$_MODULE['<{barzahlen}prestashop>barzahlen_4646a4a094a491d38e8b03b2b19a7162'] = 'Barzahlen';
$_MODULE['<{barzahlen}prestashop>barzahlen_ec8518f811230bdc887bd13c8bd47f4a'] = 'Barzahlen bietet Ihren Kunden die Möglichkeit, online bar zu bezahlen. Sie werden in Echtzeit über die Zahlung benachrichtigt und profitieren von voller Zahlungsgarantie und neuen Kundengruppen. Sehen Sie wie Barzahlen funktioniert: <a href=http://www.barzahlen.de/partner/funktionsweise target="_blank">http://www.barzahlen.de/partner/funktionsweise</a>';
$_MODULE['<{barzahlen}prestashop>barzahlen_419ad563a41b36b858d6ff0fd8b34d2a'] = 'Barzahlen Einstellungen';
$_MODULE['<{barzahlen}prestashop>barzahlen_2652eec977dcb2a5aea85f5bec235b05'] = 'Sandbox';
$_MODULE['<{barzahlen}prestashop>barzahlen_ccf107a5d46c6501c9f2f4345400dc2e'] = 'Shop ID';
$_MODULE['<{barzahlen}prestashop>barzahlen_f7a0ec6eface594351e1a6e05af59a7e'] = 'Zahlungsschlüssel';
$_MODULE['<{barzahlen}prestashop>barzahlen_ac2e21a9827e093b9955c6cb97007dca'] = 'Benachrichtigungsschlüssel';
$_MODULE['<{barzahlen}prestashop>barzahlen_24397caa1f4a3c44ed8f2da325852017'] = 'Erweitertes Logging';
$_MODULE['<{barzahlen}prestashop>barzahlen_b17f3f4dcf653a5776792498a9b44d6a'] = 'Einstellungen speichern';
$_MODULE['<{barzahlen}prestashop>barzahlen_c5bbe1f842cfd421b2e601dc37291b48'] = 'Für das Barzahlen-Plugin ist eine neue Version (%s) verfügbar. Jetzt ansehen?';
$_MODULE['<{barzahlen}prestashop>payment_75348b9b25745cf384d22b99b647a9d4'] = 'Mit Barzahlen bezahlen';
$_MODULE['<{barzahlen}prestashop>payment_4646a4a094a491d38e8b03b2b19a7162'] = 'Barzahlen';
$_MODULE['<{barzahlen}prestashop>payment_5abc9461c112d4075023c0d26370c02a'] = 'Bezahlen Sie bei:';
$_MODULE['<{barzahlen}prestashop>payment_execution_4646a4a094a491d38e8b03b2b19a7162'] = 'Barzahlen';
$_MODULE['<{barzahlen}prestashop>payment_execution_f1d3b424cd68795ecaa552883759aceb'] = 'Bestellübersicht';
$_MODULE['<{barzahlen}prestashop>payment_execution_879f6b8877752685a966564d072f498f'] = 'Ihr Warenkorb ist leer.';
$_MODULE['<{barzahlen}prestashop>payment_execution_fa383bfb4ab05b2e3edf2212718ea267'] = 'Mit Abschluss der Bestellung bekommen Sie einen Zahlschein angezeigt, den Sie sich ausdrucken oder auf Ihr Handy schicken lassen können. Bezahlen Sie den Online-Einkauf mit Hilfe des Zahlscheins an der Kasse einer Barzahlen-Partnerfiliale.';
$_MODULE['<{barzahlen}prestashop>payment_execution_361eac23da63287072af87c6abda8474'] = 'Der Sandbox Modus ist aktiv. Allen getätigten Zahlungen wird ein Test-Zahlschein zugewiesen. Dieser kann nicht von unseren Einzelhandelspartnern verarbeitet werden.';
$_MODULE['<{barzahlen}prestashop>payment_execution_5abc9461c112d4075023c0d26370c02a'] = 'Bezahlen Sie bei:';
$_MODULE['<{barzahlen}prestashop>payment_execution_46b9e3665f187c739c55983f757ccda0'] = 'Kostenpflichtig bestellen';
$_MODULE['<{barzahlen}prestashop>payment_execution_569fd05bdafa1712c4f6be5b153b8418'] = 'Andere Zahlungsmethoden';


src/modules/barzahlen/api/api.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

class Barzahlen_Api extends Barzahlen_Base {

  protected $_shopId; //!< merchants shop id
  protected $_paymentKey; //!< merchants payment key
  protected $_language = 'de'; //!< langauge code
  protected $_allowLanguages = array('de', 'en'); //!< allowed languages for requests
  protected $_sandbox = false; //!< sandbox settings
  protected $_madeAttempts = 0; //!< performed attempts

  /**
   * Constructor. Sets basic settings.
   *
   * @param string $shopId merchants shop id
   * @param string $paymentKey merchants payment key
   * @param boolean $sandbox sandbox settings (default: false -> off)
   */
  public function __construct($shopId, $paymentKey, $sandbox = false) {

    $this->_shopId = $shopId;
    $this->_paymentKey = $paymentKey;
    $this->_sandbox = $sandbox;
  }

  /**
   * Sets the language for payment / refund slip.
   *
   * @param string $language Langauge Code (ISO 639-1)
   */
  public function setLanguage($language = 'de') {

    if(in_array($language, $this->_allowLanguages)) {
      $this->_language = $language;
    }
    else {
      $this->_language = $this->_allowLanguages[0];
    }
  }

  /**
   * Handles request of all kinds.
   *
   * @param Barzahlen_Request $request request that should be made
   */
  public function handleRequest($request) {

    $requestArray = $request->buildRequestArray($this->_shopId, $this->_paymentKey, $this->_language);
    $this->_debug("Barzahlen/API: Sending request array to Barzahlen.", $requestArray);
    $xmlResponse = $this->_connectToApi($requestArray, $request->getRequestType());
    $this->_debug("Barzahlen/API: Received XML response from Barzahlen.");
    $request->parseXml($xmlResponse, $this->_paymentKey);
    $this->_debug("Barzahlen/API: Parsed XML response and returned it to request object.");
  }

  /**
   * Connects to Barzahlen Api as long as there's a xml response or maximum attempts are reached.
   *
   * @param array $requestArray array with the information which shall be send via POST
   * @param string $requestType type for request
   * @return xml response from Barzahlen
   */
  protected function _connectToApi(array $requestArray, $requestType) {

    $this->_madeAttempts++;
    $curl = $this->_prepareRequest($requestArray, $requestType);

    try {
      return $this->_sendRequest($curl);
    }
    catch (Exception $e) {
      if ($this->_madeAttempts >= self::MAXATTEMPTS) {
        throw new Barzahlen_Exception ($e->getMessage());
      }
      return $this->_connectToApi($requestArray, $requestType);
    }
  }

  /**
   * Prepares the curl request.
   *
   * @param array $requestArray array with the information which shall be send via POST
   * @param string $requestType type of request
   * @return cURL handle object
   */
  protected function _prepareRequest(array $requestArray, $requestType) {

    $callDomain = $this->_sandbox ? self::APIDOMAINSANDBOX.$requestType : self::APIDOMAIN.$requestType;

    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $callDomain);
    curl_setopt($curl, CURLOPT_POST, count($requestArray));
    curl_setopt($curl, CURLOPT_POSTFIELDS, $requestArray);
    curl_setopt($curl, CURLOPT_HEADER, 0);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, true);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, 2);
    curl_setopt($curl, CURLOPT_CAINFO, dirname(__FILE__) . '/certs/ca-bundle.crt');
    curl_setopt($curl, CURLOPT_TIMEOUT, 30);
    curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 15);
    curl_setopt($curl, CURLOPT_HTTP_VERSION, 1.1);
    return $curl;
  }

  /**
   * Send the information via HTTP POST to the given domain. A xml as anwser is expected.
   * SSL is required for a connection to Barzahlen.
   *
   * @return cURL handle object
   * @return xml response from Barzahlen
   */
  protected function _sendRequest($curl) {

    $return = curl_exec($curl);
    $error = curl_error($curl);
    curl_close($curl);

    if($error != '') {
      throw new Barzahlen_Exception('Error during cURL: ' . $error);
    }

    return $return;
  }
}
?>

src/modules/barzahlen/payment.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

/* SSL Management */
$useSSL = true;

require_once dirname(__FILE__) . '/../../config/config.inc.php';
require_once dirname(__FILE__) . '/../../header.php';
require_once dirname(__FILE__) . '/barzahlen.php';

if (!$cookie->isLogged(true)) {
    Tools::redirect('authentication.php?back=order.php');
}

$barzahlen = new Barzahlen();
echo $barzahlen->execPayment($cart);

require_once dirname(__FILE__) . '/../../footer.php';


src/modules/barzahlen/api/base.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

abstract class Barzahlen_Base {

  const APIDOMAIN = 'https://api.barzahlen.de/v1/transactions/'; //!< call domain (productive use)
  const APIDOMAINSANDBOX = 'https://api-sandbox.barzahlen.de/v1/transactions/'; //!< sandbox call domain

  const HASHALGO = 'sha512'; //!< hash algorithm
  const SEPARATOR = ';'; //!< separator character
  const MAXATTEMPTS = 2; //!< maximum of allowed connection attempts

  protected $_debug = false; //!< debug mode on / off

  /**
   * Sets debug settings. Adjusted for PrestaShop.
   *
   * @param boolean $debug debug mode on / off
   * @param string $logFile position of log file
   */
  public function setDebug($debug) {
    $this->_debug = $debug;
  }

  /**
   * Write debug message to log file. Adjusted for PrestaShop.
   *
   * @param string $message debug message
   * @param array $data related data (optional)
   */
  protected function _debug($message, $data = array()) {

    if($this->_debug == true) {
      foreach ($data as $key => $value) {
        $message .= ' - ' . $key . ':' . $value;
      }
      LoggerCore::addLog($message, 1, null, null, null, true);
    }
  }

  /**
   * Generates the hash for the request array.
   *
   * @param array $requestArray array from which hash is requested
   * @param string $key private key depending on hash type
   * @return hash sum
   */
  protected function _createHash(array $hashArray, $key) {

    $hashArray[] = $key;
    $hashString = implode(self::SEPARATOR, $hashArray);
    return hash(self::HASHALGO, $hashString);
  }

  /**
   * Removes empty values from arrays.
   *
   * @param array $array array with (empty) values
   */
  protected function _removeEmptyValues(array &$array) {

    foreach($array as $key => $value) {
      if($value == '') {
        unset($array[$key]);
      }
    }
  }

  /**
   * Converts ISO-8859-1 strings to UTF-8 if necessary.
   *
   * @param string $string text which is to check
   * @return string with utf-8 encoding
   */
  public function isoConvert($string) {

    if(!preg_match('/\S/u', $string)) {
      $string = utf8_encode($string);
    }

    return $string;
  }
}
?>

src/modules/barzahlen/callback.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

require_once dirname(__FILE__) . '/../../config/config.inc.php';
require_once dirname(__FILE__) . '/api/loader.php';

// let's roll
new BarzahlenCallback;

class BarzahlenCallback
{
    const STATE_PENDING = 'pending';
    const STATE_PAID = 'paid';
    const STATE_EXPIRED = 'expired';

    /**
     * @see FrontController::initContent()
     */
    public function __construct()
    {
        $notification = new Barzahlen_Notification(Configuration::get('barzahlen_shopid'), Configuration::get('barzahlen_notificationkey'), $_GET);

        try {
            $notification->validate();
        } catch (Exception $e) {
            LoggerCore::addLog('Barzahlen/Payment: ' . $e, 3, null, null, null, true);
        }

        if (!$notification->isValid()) {
            $this->_sendHeader(400);
        }

        $this->_sendHeader(200);
        $result = $this->_selectTransaction($notification);

        if (count($result) == 0) {
            LoggerCore::addLog('Barzahlen/Callback: No pending transaction found for order ID ' . $notification->getOrderId() . ' and transaction ID ' . $notification->getTransactionId() . '.', 3, null, null, null, true);
            return;
        } else {
            $order = new Order($result[0]['order_id']);
        }

        switch ($notification->getState()) {
            case self::STATE_PAID:
                $order->setCurrentState(Configuration::get('BARZAHLEN_PAID'));
                $this->_updateTransactionState($notification->getTransactionId(), self::STATE_PAID);
                break;
            case self::STATE_EXPIRED:
                $order->setCurrentState(Configuration::get('BARZAHLEN_EXPIRED'));
                $this->_updateTransactionState($notification->getTransactionId(), self::STATE_EXPIRED);
                break;
            default:
                LoggerCore::addLog('Barzahlen/Callback: Unable to handle given state ' . $notification->getState() . '.', 3, null, null, null, true);
        }
    }

    /**
     * Looks up the corresponding from the databse.
     *
     * @param Barzahlen_Notification $notification
     * @return array
     */
    protected function _selectTransaction(Barzahlen_Notification $notification)
    {
        $sql = "SELECT * FROM `" . _DB_PREFIX_ . "barzahlen_transactions`
                 WHERE transaction_id = '" . (int) $notification->getTransactionId() . "'
                   AND transaction_state = '" . self::STATE_PENDING . "'";

        if ($notification->getOrderId() != null) {
            $sql .= " AND order_id = '" . (int) $notification->getOrderId() . "'";
        }

        return Db::getInstance()->executeS($sql);
    }

    /**
     * Updates the transaction in the database.
     *
     * @param integer $transactionId transaction id
     * @param string $state new transaction state
     */
    protected function _updateTransactionState($transactionId, $state)
    {
        $sql = "UPDATE `" . _DB_PREFIX_ . "barzahlen_transactions`
                   SET transaction_state = '" . $state . "'
                 WHERE transaction_id = '" . (int) $transactionId . "'";

        Db::getInstance()->execute($sql);
    }

    /**
     * Send the header depending on the notification validation.
     *
     * @param integer $code status code
     */
    protected function _sendHeader($code)
    {
        if ($code == 200) {
            header("HTTP/1.1 200 OK");
            header("Status: 200 OK");
        } else {
            header("HTTP/1.1 400 Bad Request");
            header("Status: 400 Bad Request");
            die();
        }
    }
}


src/modules/barzahlen/barzahlen.php
<?php
/**
 * Barzahlen Payment Module (PrestaShop)
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to info@barzahlen.de so we can send you a copy immediately.
 *
 * @copyright   Copyright (c) 2012 Zerebro Internet GmbH (http://www.barzahlen.de)
 * @author      Alexander Diebler
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL-3.0)
 */

if (!defined('_PS_VERSION_')) {
    die();
}

class Barzahlen extends PaymentModule
{
    private $_html = '';
    protected $_sandbox = false;
    protected $_shopid = '';
    protected $_paymentkey = '';
    protected $_notificationkey = '';
    protected $_debug = false;

    /**
     * Constructor is used to load all necessary payment module information.
     */
    public function __construct()
    {
        $this->name = 'barzahlen';
        $this->tab = 'payments_gateways';
        $this->version = '1.0.2';
        $this->author = 'Zerebro Internet GmbH';
        $this->currencies = true;
        $this->currencies_mode = 'checkbox';

        $config = Configuration::getMultiple(array($this->name . '_sandbox', $this->name . '_shopid', $this->name . '_paymentkey', $this->name . '_notificationkey', $this->name . '_debug'));
        if (isset($config[$this->name . '_sandbox']))
            $this->_sandbox = $config[$this->name . '_sandbox'] == 'on' ? true : false;
        if (isset($config[$this->name . '_shopid']))
            $this->_shopid = $config[$this->name . '_shopid'];
        if (isset($config[$this->name . '_paymentkey']))
            $this->_paymentkey = $config[$this->name . '_paymentkey'];
        if (isset($config[$this->name . '_notificationkey']))
            $this->_notificationkey = $config[$this->name . '_notificationkey'];
        if (isset($config[$this->name . '_debug']))
            $this->_debug = $config[$this->name . '_debug'] == 'on' ? true : false;

        parent::__construct();

        $this->page = basename(__FILE__, '.php');
        $this->displayName = $this->l('Barzahlen');
        $this->description = $this->l('Barzahlen let\'s your customers pay cash online. You get a payment confirmation in real-time and you benefit from our payment guarantee and new customer groups. See how Barzahlen works: <a href=http://www.barzahlen.de/partner/funktionsweise target="_blank">http://www.barzahlen.de/partner/funktionsweise</a>');
    }

    /**
     * Core install method.
     *
     * @return boolean
     */
    public function install()
    {
        if (!parent::install() || !$this->registerHook('payment') || !$this->registerHook('paymentReturn') || !$this->registerHook('updateOrderStatus') || !$this->registerHook('backOfficeFooter')) {
            return false;
        }
        $this->createTables();
        $this->createOrderState();
        return true;
    }

    /**
     * Creates the Barzahlen transaction table.
     */
    protected function createTables()
    {
        Db::getInstance()->Execute("
            CREATE TABLE IF NOT EXISTS `" . _DB_PREFIX_ . "barzahlen_transactions` (
            `transaction_id` int(11) NOT NULL DEFAULT 0,
            `order_id` int(11) NOT NULL,
            `transaction_state` ENUM( 'pending', 'paid', 'expired' ) NOT NULL,
            PRIMARY KEY (`transaction_id`)
          ) ENGINE=" . _MYSQL_ENGINE_ . " DEFAULT CHARSET=utf8");

        // update 1.0.2
        Db::getInstance()->Execute("
            ALTER TABLE `" . _DB_PREFIX_ . "barzahlen_transactions` CHANGE `transaction_state` `transaction_state` ENUM( 'pending', 'paid', 'expired', 'canceled' ) NOT NULL;");
    }

    /**
     * Creates the new order states for the Barzahlen transaction states.
     */
    protected function createOrderState()
    {
        // pending state
        if (!Configuration::get('BARZAHLEN_PENDING')) {
            $orderState = new OrderState();
            $orderState->name = array();

            foreach (Language::getLanguages() as $language) {
                if (strtolower($language['iso_code']) == 'de') {
                    $orderState->name[$language['id_lang']] = 'Warten auf Zahlungseingang von Barzahlen';
                } else {
                    $orderState->name[$language['id_lang']] = 'Awaiting Barzahlen Payment';
                }
            }

            $orderState->send_email = false;
            $orderState->color = '#ADD8E6';
            $orderState->hidden = false;
            $orderState->delivery = false;
            $orderState->logable = true;
            $orderState->invoice = false;

            if ($orderState->add()) {
                $source = dirname(__FILE__) . '/../../img/admin/gold.gif';
                $destination = dirname(__FILE__) . '/../../img/os/' . (int) $orderState->id . '.gif';
                copy($source, $destination);
            }
            Configuration::updateValue('BARZAHLEN_PENDING', (int) $orderState->id);
        }

        // paid state
        if (!Configuration::get('BARZAHLEN_PAID')) {
            $orderState = new OrderState();
            $orderState->name = array();

            foreach (Language::getLanguages() as $language) {
                if (strtolower($language['iso_code']) == 'de') {
                    $orderState->name[$language['id_lang']] = 'Zahlungseingang von Barzahlen';
                } else {
                    $orderState->name[$language['id_lang']] = 'Received Barzahlen Payment';
                }
            }

            $orderState->send_email = false;
            $orderState->color = '#DDEEFF';
            $orderState->hidden = false;
            $orderState->delivery = true;
            $orderState->logable = true;
            $orderState->invoice = true;

            if ($orderState->add()) {
                $source = dirname(__FILE__) . '/../../img/os/2.gif';
                $destination = dirname(__FILE__) . '/../../img/os/' . (int) $orderState->id . '.gif';
                copy($source, $destination);
            }
            Configuration::updateValue('BARZAHLEN_PAID', (int) $orderState->id);
        }

        // expired state
        if (!Configuration::get('BARZAHLEN_EXPIRED')) {
            $orderState = new OrderState();
            $orderState->name = array();

            foreach (Language::getLanguages() as $language) {
                if (strtolower($language['iso_code']) == 'de') {
                    $orderState->name[$language['id_lang']] = 'Barzahlen-Zahlschein abgelaufen';
                } else {
                    $orderState->name[$language['id_lang']] = 'Barzahlen Payment Expired';
                }
            }

            $orderState->send_email = false;
            $orderState->color = '#FFDFDF';
            $orderState->hidden = false;
            $orderState->delivery = false;
            $orderState->logable = true;
            $orderState->invoice = false;

            if ($orderState->add()) {
                $source = dirname(__FILE__) . '/../../img/os/6.gif';
                $destination = dirname(__FILE__) . '/../../img/os/' . (int) $orderState->id . '.gif';
                copy($source, $destination);
            }
            Configuration::updateValue('BARZAHLEN_EXPIRED', (int) $orderState->id);
        }
    }

    /**
     * Uninstaller. Extends parent and removes Barzahlen settings. Transaction
     * table and order states remain.
     *
     * @return boolean
     */
    public function uninstall()
    {
        if (!Configuration::deleteByName($this->name . '_sandbox')
                || !Configuration::deleteByName($this->name . '_shopid')
                || !Configuration::deleteByName($this->name . '_paymentkey')
                || !Configuration::deleteByName($this->name . '_notificationkey')
                || !Configuration::deleteByName($this->name . '_debug')
                || !parent::uninstall()) {
            return false;
        }
        return true;
    }

    /**
     * Saves new settings and calls html output method.
     *
     * @return string with html code
     */
    public function getContent()
    {
        if (Tools::isSubmit('btnSubmit')) {
            Configuration::updateValue($this->name . '_sandbox', Tools::getValue($this->name . '_sandbox'));
            Configuration::updateValue($this->name . '_shopid', Tools::getValue($this->name . '_shopid'));
            Configuration::updateValue($this->name . '_paymentkey', Tools::getValue($this->name . '_paymentkey'));
            Configuration::updateValue($this->name . '_notificationkey', Tools::getValue($this->name . '_notificationkey'));
            Configuration::updateValue($this->name . '_debug', Tools::getValue($this->name . '_debug'));
            $this->_sandbox = Tools::getValue($this->name . '_sandbox');
            $this->_debug = Tools::getValue($this->name . '_debug');
        }
        $this->_displayForm();
        return $this->_html;
    }

    /**
     * Prepares the html form for the module configuration.
     */
    private function _displayForm()
    {
        $sandboxChecked = $this->_sandbox ? 'checked="checked"' : '';
        $debugChecked = $this->_debug ? 'checked="checked"' : '';

        $this->_html .=
                '<form action="' . Tools::htmlentitiesUTF8($_SERVER['REQUEST_URI']) . '" method="post">
      <fieldset>
      <legend><img src="../img/admin/prefs.gif" />' . $this->l('Barzahlen Settings') . '</legend>
        <table border="0" width="500" cellpadding="0" cellspacing="0" id="form">
          <tr><td width="170" style="height: 35px;">' . $this->l('Sandbox') . '</td><td><input type="checkbox" name="' . $this->name . '_sandbox" ' . $sandboxChecked . '/></td></tr>
          <tr><td width="170" style="height: 35px;">' . $this->l('Shop ID') . '</td><td><input type="text" name="' . $this->name . '_shopid" value="' . htmlentities(Tools::getValue($this->name . '_shopid', $this->_shopid), ENT_COMPAT, 'UTF-8') . '" style="width: 300px;" /></td></tr>
          <tr><td width="170" style="height: 35px;">' . $this->l('Payment Key') . '</td><td><input type="text" name="' . $this->name . '_paymentkey" value="' . htmlentities(Tools::getValue($this->name . '_paymentkey', $this->_paymentkey), ENT_COMPAT, 'UTF-8') . '" style="width: 300px;" /></td></tr>
          <tr><td width="170" style="height: 35px;">' . $this->l('Notification Key') . '</td><td><input type="text" name="' . $this->name . '_notificationkey" value="' . htmlentities(Tools::getValue($this->name . '_notificationkey', $this->_notificationkey), ENT_COMPAT, 'UTF-8') . '" style="width: 300px;" /></td></tr>
          <tr><td width="170" style="height: 35px;">' . $this->l('Extended Logging') . '</td><td><input type="checkbox" name="' . $this->name . '_debug" ' . $debugChecked . '/></td></tr>
          <tr><td colspan="2" align="center"><input class="button" name="btnSubmit" value="' . $this->l('Update settings') . '" type="submit" /></td></tr>
        </table>
      </fieldset>
    </form>';
    }

    /**
     * Prepares and returns payment template for payment selection hook.
     *
     * @global smarty object $smarty
     * @param array $params order parameters
     * @return boolean | string rendered template output
     */
    public function hookPayment($params)
    {
        global $smarty;

        if (!$this->active || !$this->_checkCurrency($params['cart'])) {
            return;
        }

        if ($params['cart']->getOrderTotal(true, Cart::BOTH) >= 1000) {
            return;
        }

        $smarty->assign(array(
            'this_path' => $this->_path,
            'this_path_ssl' => Tools::getShopDomainSsl(true, true) . __PS_BASE_URI__ . 'modules/' . $this->name . '/',
            $this->name . '_sandbox' => $this->_sandbox
        ));
        return $this->display(__FILE__, 'payment.tpl');
    }

    /**
     * Hook for catch order cancelations to cancel pending payment slips parallel.
     *
     * @param type $params
     */
    public function hookUpdateOrderStatus($params)
    {
        if ($params['newOrderStatus']->template != 'order_canceled') {
            return;
        }

        $rs = Db::getInstance()->ExecuteS("SELECT transaction_id, transaction_state FROM `" . _DB_PREFIX_ . "barzahlen_transactions` WHERE order_id = '" . (int) $params['id_order'] . "'");

        if ($rs[0]['transaction_state'] != 'pending') {
            return;
        }

        global $cookie;
        require_once dirname(__FILE__) . '/api/loader.php';

        $lang = new Language((int) $cookie->id_lang);
        $api = new Barzahlen_Api(Configuration::get('barzahlen_shopid'), Configuration::get('barzahlen_paymentkey'), Configuration::get('barzahlen_sandbox'));
        $api->setDebug(Configuration::get('barzahlen_debug'));
        $api->setLanguage($lang->iso_code);

        $cancel = new Barzahlen_Request_Cancel($rs[0]['transaction_id']);

        try {
            $api->handleRequest($cancel);
        } catch (Exception $e) {
            LoggerCore::addLog('Barzahlen/Cancel: ' . $e, 3, null, null, null, true);
        }

        if ($cancel->isValid()) {
            Db::getInstance()->Execute("UPDATE `" . _DB_PREFIX_ . "barzahlen_transactions` SET transaction_state = 'canceled' WHERE transaction_id = '" . $rs[0]['transaction_id'] . "'");
        }
    }

    /**
     * Automatic plugin version check (once a week).
     *
     * @param type $params
     */
    public function hookBackOfficeFooter($params)
    {
        $lastCheck = Configuration::get('barzahlen_lastcheck');

        if($lastCheck == null || $lastCheck < strtotime("-1 week")) {

            Configuration::updateValue($this->name . '_lastcheck', time());
            require_once dirname(__FILE__) . '/api/loader.php';

            try {
                $checker = new Barzahlen_Version_Check(Configuration::get('barzahlen_shopid'), Configuration::get('barzahlen_paymentkey'));
                $response = $checker->checkVersion('PrestaShop', Configuration::get('PS_INSTALL_VERSION'), $this->version);
                if ($response != false) {
                    echo '<script type="text/javascript">
                          if (confirm("' . sprintf($this->l('For the Barzahlen plugin is a new version (%s) available. View now?'), (string) $response) . '")) {
                              window.location.href = "http://www.barzahlen.de/partner/integration/shopsysteme/14/prestashop";
                          }
                          </script>';
                }
            } catch (Exception $e) {
                LoggerCore::addLog('Barzahlen/Check: ' . $e, 3, null, null, null, true);
            }
        }
    }

    /**
     * Prepares and returns payment success template after order completion.
     *
     * @global type $smarty
     * @param type $params
     * @return string rendered template output
     */
    public function hookPaymentReturn($params)
    {
        global $smarty;

        if (!$this->active) {
            return;
        }

        session_start();
        $smarty->assign($this->name . '_infotext', $_SESSION['barzahlen_infotext']);
        unset($_SESSION['barzahlen_infotext']);

        return $this->display(__FILE__, 'confirmation.tpl');
    }

    /**
     * Get execute payment HTML.
     *
     * @global object $smarty smarty handler
     * @param object $cart customer cart
     * @return string
     */
    public function execPayment($cart)
    {
        global $smarty;

        if (!$this->active || !$this->_checkCurrency($cart)) {
            return;
        }

        $smarty->assign(array(
            'nbProducts' => $cart->nbProducts(),
            'isoCode' => $this->context->language->iso_code,
            'barzahlen_sandbox' => Configuration::get('barzahlen_sandbox'),
            'this_path' => $this->_path,
            'this_path_ssl' => Tools::getShopDomainSsl(true, true) . __PS_BASE_URI__ . 'modules/' . $this->name . '/',
        ));

        return $this->display(__FILE__, 'payment_execution.tpl');
    }

    /**
     * Checks if selected currency is possible with Barzahlen.
     *
     * @param Cart $cart cart object
     * @return boolean
     */
    protected function _checkCurrency($cart)
    {
        $currency_order = new Currency($cart->id_currency);
        $currencies_module = $this->getCurrency($cart->id_currency);

        if (is_array($currencies_module)) {
            foreach ($currencies_module as $currency_module) {
                if ($currency_order->id == $currency_module['id_currency']) {
                    return true;
                }
            }
        }
        return false;
    }
}
